### ISSUE TITLE
`TypeVar` bound to `Union` in `if isinstance(...)` branch has incompatible return type

### ISSUE THREAD

**Issue by headtr1ck** [python/mypy#12800]
**Bug Report**

A TypeVar bound to `Union[x, y]` will resolve to `x` after an `if isinstance(..., x)` is used which is incompatible with the return type.

I'm not sure on how to describe exactly what is going on, better check the Example.
Also mypy experts, please forgive me if that is intended behavior and not a bug (I would appreciate a working solution in this case :)

**To Reproduce**

Example:
```python
from typing import TypeVar, Union

T = TypeVar("T", bound=Union[str, int])

def a(x: T) -> T:
    if isinstance(x, str):
        return x
    return x
```

**Expected Behavior**

I expect that this definition is correct, since whatever input is given, the output is the same.

**Actual Behavior**

mypy raises `error: Incompatible return value type (got "str", expected "T")`.

**Further Info**

* I want to have the `str -> str`, `int -> int` mapping, so simply using `Union[str, int]` won't work.
* This function is called sometimes with variables of type `Union[str, int]`, so using `TypeVar("T", str, int)` won't work either.

**Your Environment**

- Mypy version used: 0.950
- Mypy command-line flags:  --strict
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: 3.9
- Operating system and version: win10

**Comment by AlexWaygood** [comment]
This one looks like a bug to me :)

For a (horrible) workaround, [you can use overloads instead](https://mypy-play.net/?mypy=latest&python=3.10&flags=strict&gist=8ff574aae792cecf5499961ec1c342cb):

```python
from typing import overload

@overload
def a(x: str) -> str: ...

@overload
def a(x: int) -> int: ...

@overload
def a(x: int | str) -> int | str: ...

def a(x: int | str) -> int | str:
    if isinstance(x, str):
        return x
    return x

b: str
c: int
d: int | str

reveal_type(a(b))  # Revealed type is "builtins.str"
reveal_type(a(c))  # Revealed type is "builtins.int"
reveal_type(a(d))  # Revealed type is "Union[builtins.int, builtins.str]"
```

**Comment by sterliakov** [comment]
I don't know what does it mean for debugging, but this is also a bit odd:

```python
from typing import TypeVar, Union

T = TypeVar("T", bound=Union[str, int])

def a(x: T) -> T:
    reveal_type(x)  # T`-1
    if isinstance(x, str):
        reveal_type(x)  # str
        return x
    reveal_type(x)  # T`-1
    return x
```

Gives ``T`-1``, `str` and ``T`-1`` revealed (in order) and error on line 8. However, the following:

```python
from typing import TypeVar, Union

T = TypeVar("T", bound=Union[str, int])

def a(x: T) -> T:
    reveal_type(x)  # T`-1
    if isinstance(x, str | None):
        reveal_type(x)  # T`-1
        return x
    reveal_type(x)  # T`-1
    return x
```

Gives ``T`-1`` three times and no error. But I just add `None` which is not a member of `T` bounds, so it shouldn't affect inference, right?.. Say, the following ignores `None` in similar case:
```python
x: str
if isinstance(x, str | None):
    reveal_type(x)  # str
```

**Comment by AlexWaygood** [comment]
@headtr1ck, a slightly simpler and less horrible workaround: [just use a `cast()`](https://mypy-play.net/?mypy=latest&python=3.10&gist=b03797af285e3b6f425e6f09c347473d):

```python
from typing import TypeVar, Union, cast

T = TypeVar("T", bound=Union[str, int])

def a(x: T) -> T:
    if isinstance(x, str):
        return cast(T, x)
    return x
```

**Comment by headtr1ck** [comment]
> @headtr1ck, a slightly simpler and less horrible workaround: [just use a `cast()`](https://mypy-play.net/?mypy=latest&python=3.10&gist=b03797af285e3b6f425e6f09c347473d):

This is what I ended up doing in the moment.

**Comment by gschaffner** [comment]
Is this a duplicate of #10817?

**Comment by hauntsaninja** [comment]
Closing as duplicate of https://github.com/python/mypy/issues/10817

**Comment by moranabadie** [comment]
> Closing as duplicate of #10817

https://github.com/python/mypy/issues/10817 is closed but the problem persists in python 1.10.1

**Comment by AlexWaygood** [comment]
> #10817 is closed

not true, it is still open
